pkg/services/ngalert/store/alert_rule.go

```diff
@@ -401,8 +401,8 @@ func (st DBstore) GetNamespaceByUID(ctx context.Context, uid string, orgID int64
 // GetAlertRulesForScheduling returns a short version of all alert rules except those that belong to an excluded list of organizations
 func (st DBstore) GetAlertRulesForScheduling(ctx context.Context, query *ngmodels.GetAlertRulesForSchedulingQuery) error {
 	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		alerts := make([]*ngmodels.SchedulableAlertRule, 0)
```

Comment:
Can you benchmark the allocations and memory before and after this change? I'm concerned about fetching the full set from the database every refresh interval.

Would it be better to change the behavior more extensively:
If there is nothing in the registry, do a "full fetch". 
If the registry is populated, fetch only the minimal set of data - keys and versions. If there's a version change, issue a second query to update all rules with a version change.

Then in the common refresh case our reads are still lightweight, and we still remove the "N queries" issue where each rule queries the database directly.

Reply:
I like your idea. I was thinking in a slightly different direction: support soft-delete and then do only incremental queries `(SELECT * FROM alert_rule WHERE updated > MAX(in_memory_updated))`. Your approach would let us achieve same effect without soft-deletes.

Reply:
yes, the reduction is not noticeable on that scale because the titles and group name of alerts rules in my test are very short.